src/permutat.cc

```diff
@@ -2649,150 +2649,109 @@ static Obj FuncMappingPermListList(Obj self, Obj src, Obj dst)
 /* end ); */
 
 
-static Obj FuncSCR_SIFT_HELPER(Obj self, Obj S, Obj g, Obj n)
-{
-  Obj stb = S;
-  static UInt RN_stabilizer = 0;
-  static UInt RN_orbit = 0;
-  static UInt RN_transversal = 0;
-  UInt nn = INT_INTOBJ(n);
-  UInt useP2;
-  Obj result;
-  Obj t;
-  Obj trans;
-  int i;
-
-  /* Setup the result, sort out which rep we are going to work in  */
-  if (nn > 65535) {
-    result = NEW_PERM4(nn);
-    useP2 = 0;
-  } else {
-    result = NEW_PERM2(nn);
-    useP2 = 1;
-  }
+static UInt RN_stabilizer = 0;
+static UInt RN_orbit = 0;
+static UInt RN_transversal = 0;
 
-  UInt dg;
-  if (IS_PERM2(g))
-    dg = DEG_PERM2(g);
-  else
-    dg = DEG_PERM4(g);
-
-  if (dg > nn) /* In this case the caller has messed up or 
-                  g just ends with a lot of fixed points which we can 
-                  ignore */
-    dg = nn;
-
-  /* Copy g into the buffer */
-  if (IS_PERM2(g) && useP2) {
-    UInt2 * ptR = ADDR_PERM2(result);
-    memcpy(ptR, CONST_ADDR_PERM2(g), 2*dg);
-    for ( i = dg; i < nn; i++)
-      ptR[i] = (UInt2)i;
-  } else if (IS_PERM4(g) && !useP2) {
-    UInt4 *ptR = ADDR_PERM4(result);
-    memcpy(ptR, CONST_ADDR_PERM4(g), 4*dg);
-    for ( i = dg; i <nn; i++)
-      ptR[i] = (UInt4)i;
-  } else if (IS_PERM2(g) && !useP2) {
-    UInt4 *ptR = ADDR_PERM4(result);
-    const UInt2 *ptG = CONST_ADDR_PERM2(g);
-    for ( i = 0; i < dg; i++)
-      ptR[i] = (UInt4)ptG[i];
-    for (i = dg; i < nn; i++)
-      ptR[i] = (UInt4)i;
-  } else {
-    UInt2 *ptR = ADDR_PERM2(result);
-    const UInt4 *ptG = CONST_ADDR_PERM4(g);
-    for ( i = 0; i < dg; i++)
-      ptR[i] = (UInt2)ptG[i];
-    for (i = dg; i < nn; i++)
-      ptR[i] = (UInt2)i;
-  }    
-    
-  
-  if (!RN_stabilizer)
-    RN_stabilizer = RNamName("stabilizer");
-  if (!RN_orbit)
-    RN_orbit = RNamName("orbit");
-  if (!RN_transversal)
-    RN_transversal = RNamName("transversal");
-  while (IsbPRec(stb,RN_stabilizer)) {
-    trans = ElmPRec(stb, RN_transversal);
-    Obj orb = ElmPRec(stb,RN_orbit);
-    Int bpt = INT_INTOBJ(ELM_LIST(orb,1))-1;
-    Int im;
-
-    if (useP2) {
-        const UInt2* ptrResult = CONST_ADDR_PERM2(result);
-        UInt degResult = DEG_PERM2(result);
-        im = (Int)(IMAGE(bpt, ptrResult, degResult));
+template <typename TG, typename Res>
+static Obj SCR_SIFT_HELPER(Obj stb, Obj g, UInt nn)
+{
+    int  i;
+    Obj result = NEW_PERM<Res>(nn);
+    UInt dg = DEG_PERM<TG>(g);
+
+    if (dg > nn) /* In this case the caller has messed up or
+                    g just ends with a lot of fixed points which we can
+                    ignore */
+        dg = nn;
+
+    /* Copy g into the buffer */
+    Res * ptR = ADDR_PERM<Res>(result);
+    const TG * ptG = CONST_ADDR_PERM<TG>(g);
+    if (sizeof(TG) == sizeof(Res)) {
+        memcpy(ptR, ptG, sizeof(TG) * dg);
     }
     else {
-        const UInt4* ptrResult = CONST_ADDR_PERM4(result);
-        UInt degResult = DEG_PERM2(result);
```

Comment:
@stevelinton this is the incorrect `DEG_PERM2` call.

Also, I wonder: isn't the value returned by DEG_PERM4 here equal to `nn` anyway?